File: lib/pacProfile.ts
import { getCache, setCache } from "./cache"
import {
  getAllFecCandidates,
  getMembersFromDb,
  getPacDonationsByName,
  type DbMember,
} from "./db"

export type PacRecipient = {
  bioguideId: string
  name: string
  party: "D" | "R" | "I"
  chamber: "house" | "senate"
  amount: number
}

export type PacRecipients = {
  pacName: string
  totalAmount: number
  houseCount: number
  senateCount: number
  recipients: PacRecipient[]
}

export type SpendingPoint = { t: number; c: number }

const RECIPIENTS_TTL_SECONDS = 6 * 60 * 60
const COMMITTEE_ID_TTL_SECONDS = 30 * 24 * 60 * 60
const COMMITTEE_MISS_TTL_SECONDS = 6 * 60 * 60
const SPENDING_TTL_SECONDS = 12 * 60 * 60
const REPORT_MAX_PAGES = 3

function normalizeParty(value: string | null | undefined): "D" | "R" | "I" {
  const v = (value ?? "").trim().toUpperCase()
  if (v.startsWith("D")) return "D"
  if (v.startsWith("R")) return "R"
  return "I"
}


export async function getPacRecipients(pacName: string): Promise<PacRecipients> {
  const cacheKey = `pac-recipients:${pacName.toLowerCase()}`
  const cached = await getCache<PacRecipients>(cacheKey)
  if (cached) return cached

  const [rows, house, senate] = await Promise.all([
    getPacDonationsByName(pacName),
    getMembersFromDb("house"),
    getMembersFromDb("senate"),
  ])

  const byId = new Map<string, DbMember>()
  for (const m of [...house, ...senate]) byId.set(m.bioguide_id, m)

  const byMember = new Map<string, PacRecipient>()
  for (const r of rows) {
    const member = byId.get(r.bioguide_id)
    if (!member) continue
    const amount = Number(r.amount)
    if (!Number.isFinite(amount) || amount <= 0) continue
    const existing = byMember.get(r.bioguide_id)
    if (existing) {
      existing.amount += amount
    } else {
      byMember.set(r.bioguide_id, {
        bioguideId: r.bioguide_id,
        name: member.name,
        party: normalizeParty(member.party),
        chamber: member.chamber === "senate" ? "senate" : "house",
        amount,
      })
    }
  }

  const recipients = [...byMember.values()]
    .map((r) => ({ ...r, amount: Math.round(r.amount) }))
    .sort((a, b) => b.amount - a.amount)

  const result: PacRecipients = {
    pacName,
    totalAmount: recipients.reduce((sum, r) => sum + r.amount, 0),
    houseCount: recipients.filter((r) => r.chamber === "house").length,
    senateCount: recipients.filter((r) => r.chamber === "senate").length,
    recipients,
  }

  if (recipients.length > 0) await setCache(cacheKey, result, RECIPIENTS_TTL_SECONDS)
  return result
}

type FecCommittee = { committee_id?: string; committee_type?: string; name?: string }
const PAC_TYPE_PRIORITY = ["Q", "N", "O", "V", "W", "U", "D"]
const NAME_STOPWORDS = new Set([
  "POLITICAL", "ACTION", "COMMITTEE", "FUND", "NATIONAL", "AMERICAN",
  "ASSOCIATION", "FEDERAL", "INC", "CORP", "CORPORATION", "COMPANY", "THE",
])

function typeRank(type: string | undefined): number {
  const i = type ? PAC_TYPE_PRIORITY.indexOf(type) : -1
  return i === -1 ? PAC_TYPE_PRIORITY.length : i
}

function significantTokens(name: string): string[] {
  return name
    .toUpperCase()
    .split(/[^A-Z0-9]+/)
    .filter((t) => t.length >= 4 && !NAME_STOPWORDS.has(t))
}
function pickCommittee(results: FecCommittee[], tokens: string[]): string | null {
  const named = results.filter((c) => c.committee_id && c.name)
  const matches =
    tokens.length === 0
      ? named
      : named.filter((c) => {
          const upper = c.name!.toUpperCase()
          return tokens.some((t) => upper.includes(t))
        })
  if (matches.length === 0) return null
  matches.sort((a, b) => typeRank(a.committee_type) - typeRank(b.committee_type))
  return matches[0].committee_id ?? null
}

async function searchCommittees(query: string, apiKey: string): Promise<FecCommittee[]> {
  const url = new URL("https://api.open.fec.gov/v1/committees/")
  url.searchParams.set("api_key", apiKey)
  url.searchParams.set("q", query)
  url.searchParams.set("per_page", "20")
  const res = await fetch(url.toString(), { next: { revalidate: 86400 } })
  if (!res.ok) return []
  const data = await res.json()
  return (data.results ?? []) as FecCommittee[]
}

async function resolvePacCommitteeId(
  pacName: string,
  apiKey: string
): Promise<string | null> {
  const cacheKey = `pac-committee-id:${pacName.toLowerCase()}`
  const cached = await getCache<{ id: string | null }>(cacheKey)
  if (cached) return cached.id
  const cleaned = pacName
    .replace(/\bPOLITICAL ACTION COMMITTEE\b/i, "")
    .replace(/\bPAC\b/i, "")
    .replace(/\s+/g, " ")
    .trim()
  const queries =
    cleaned && cleaned.toUpperCase() !== pacName.toUpperCase()
      ? [pacName, cleaned]
      : [pacName]
  const tokens = significantTokens(pacName)

  let id: string | null = null
  try {
    for (const query of queries) {
      const results = await searchCommittees(query, apiKey)
      id = pickCommittee(results, tokens)
      if (id) break
    }
  } catch (error) {
    console.warn(`[pac] resolvePacCommitteeId(${pacName}) failed:`, error)
  }

  await setCache(cacheKey, { id }, id ? COMMITTEE_ID_TTL_SECONDS : COMMITTEE_MISS_TTL_SECONDS)
  return id
}

type FecReport = {
  coverage_end_date?: string
  total_disbursements_period?: number | string
  most_recent?: boolean
  receipt_date?: string
}

async function fetchPacSpendingHistory(
  committeeId: string,
  apiKey: string
): Promise<SpendingPoint[]> {
  const cacheKey = `pac-spending:${committeeId}`
  const cached = await getCache<SpendingPoint[]>(cacheKey)
  if (cached) return cached

  const reports: FecReport[] = []
  for (let page = 1; page <= REPORT_MAX_PAGES; page++) {
    const url = new URL(
      `https://api.open.fec.gov/v1/committee/${committeeId}/reports/`
    )
    url.searchParams.set("api_key", apiKey)
    url.searchParams.set("per_page", "100")
    url.searchParams.set("page", String(page))
    url.searchParams.set("sort", "-coverage_end_date")

    const res = await fetch(url.toString(), { next: { revalidate: 3600 } })
    if (!res.ok) break
    const data = await res.json()
    const results: FecReport[] = data.results ?? []
    if (results.length === 0) break
    reports.push(...results)
    const pages = data.pagination?.pages ?? 1
    if (page >= pages) break
  }

  const byPeriod = new Map<string, FecReport>()
  for (const r of reports) {
    if (!r.coverage_end_date || r.most_recent === false) continue
    const day = r.coverage_end_date.slice(0, 10)
    const existing = byPeriod.get(day)
    if (!existing || (r.receipt_date ?? "") > (existing.receipt_date ?? "")) {
      byPeriod.set(day, r)
    }
  }

  const points = [...byPeriod.entries()]
    .map(([day, r]) => ({
      t: Date.parse(day),
      c: Math.max(0, Number(r.total_disbursements_period ?? 0)),
    }))
    .filter((p) => Number.isFinite(p.t))
    .sort((a, b) => a.t - b.t)

  if (points.length > 0) await setCache(cacheKey, points, SPENDING_TTL_SECONDS)
  return points
}

export async function getPacSpending(
  pacName: string
): Promise<{ committeeId: string | null; points: SpendingPoint[] }> {
  const apiKey = process.env.FEC_API_KEY
  if (!apiKey) return { committeeId: null, points: [] }

  const committeeId = await resolvePacCommitteeId(pacName, apiKey)
  if (!committeeId) return { committeeId: null, points: [] }

  const points = await fetchPacSpendingHistory(committeeId, apiKey)
  return { committeeId, points }
}
export type FeedMember = {
  bioguideId: string
  name: string
  party: "D" | "R" | "I"
  chamber: "house" | "senate"
}

export type WindowContribution = { bioguideId: string; amount: number; date: number }

export type PacContributionFeed = {
  committeeId: string | null
  members: FeedMember[]
  contributions: WindowContribution[]
}

const FEED_TTL_SECONDS = 12 * 60 * 60
const FEED_MAX_PAGES = 15
const FEED_LOOKBACK_MS = 5 * 365 * 24 * 60 * 60 * 1000

type FecScheduleA = {
  committee_id?: string
  contribution_receipt_date?: string
  contribution_receipt_amount?: number | string
  committee?: { candidate_ids?: string[] }
}

type RawContribution = {
  recipientCommitteeId: string | null
  candidateIds: string[]
  date: string
  amount: number
}
async function fetchPacContributions(
  committeeId: string,
  apiKey: string
): Promise<RawContribution[]> {
  const minDate = new Date(Date.now() - FEED_LOOKBACK_MS).toISOString().slice(0, 10)
  const out: RawContribution[] = []
  let lastIndex: string | undefined
  let lastDate: string | undefined

  for (let page = 0; page < FEED_MAX_PAGES; page++) {
    const url = new URL("https://api.open.fec.gov/v1/schedules/schedule_a/")
    url.searchParams.set("api_key", apiKey)
    url.searchParams.set("contributor_id", committeeId)
    url.searchParams.set("min_date", minDate)
    url.searchParams.set("per_page", "100")
    url.searchParams.set("sort", "-contribution_receipt_date")
    if (lastIndex) url.searchParams.set("last_index", lastIndex)
    if (lastDate) url.searchParams.set("last_contribution_receipt_date", lastDate)

    const res = await fetch(url.toString(), { next: { revalidate: 3600 } })
    if (!res.ok) break
    const data = await res.json()
    const results: FecScheduleA[] = data.results ?? []
    if (results.length === 0) break

    for (const r of results) {
      const amount = Number(r.contribution_receipt_amount ?? 0)
      if (!Number.isFinite(amount) || amount <= 0) continue
      out.push({
        recipientCommitteeId: r.committee_id ?? null,
        candidateIds: r.committee?.candidate_ids ?? [],
        date: r.contribution_receipt_date ?? "",
        amount,
      })
    }

    const li = data.pagination?.last_indexes
    if (!li?.last_index || li.last_index === lastIndex) break
    lastIndex = li.last_index
    lastDate = li.last_contribution_receipt_date
  }

  return out
}
export async function getPacContributionFeed(
  pacName: string
): Promise<PacContributionFeed> {
  const apiKey = process.env.FEC_API_KEY
  if (!apiKey) return { committeeId: null, members: [], contributions: [] }

  const committeeId = await resolvePacCommitteeId(pacName, apiKey)
  if (!committeeId) return { committeeId: null, members: [], contributions: [] }

  const cacheKey = `pac-feed:${committeeId}`
  const cached = await getCache<PacContributionFeed>(cacheKey)
  if (cached) return cached

  const [raw, candidates, house, senate] = await Promise.all([
    fetchPacContributions(committeeId, apiKey),
    getAllFecCandidates(),
    getMembersFromDb("house"),
    getMembersFromDb("senate"),
  ])

  const committeeToBioguide = new Map<string, string>()
  const candidateToBioguide = new Map<string, string>()
  for (const c of candidates) {
    for (const cid of c.committee_ids ?? []) committeeToBioguide.set(cid, c.bioguide_id)
    if (c.candidate_id) candidateToBioguide.set(c.candidate_id, c.bioguide_id)
  }

  const memberById = new Map<string, DbMember>()
  for (const m of [...house, ...senate]) memberById.set(m.bioguide_id, m)

  const contributions: WindowContribution[] = []
  const present = new Map<string, FeedMember>()
  for (const r of raw) {
    let bioguide = r.recipientCommitteeId
      ? committeeToBioguide.get(r.recipientCommitteeId)
      : undefined
    if (!bioguide) {
      for (const cand of r.candidateIds) {
        const b = candidateToBioguide.get(cand)
        if (b) {
          bioguide = b
          break
        }
      }
    }
    if (!bioguide) continue
    const member = memberById.get(bioguide)
    if (!member) continue
    const date = Date.parse(r.date)
    if (!Number.isFinite(date)) continue

    contributions.push({ bioguideId: bioguide, amount: Math.round(r.amount), date })
    if (!present.has(bioguide)) {
      present.set(bioguide, {
        bioguideId: bioguide,
        name: member.name,
        party: normalizeParty(member.party),
        chamber: member.chamber === "senate" ? "senate" : "house",
      })
    }
  }

  const result: PacContributionFeed = {
    committeeId,
    members: [...present.values()],
    contributions,
  }
  if (contributions.length > 0) await setCache(cacheKey, result, FEED_TTL_SECONDS)
  return result
}

File: backend/app/services/pac_profile.py
"""PAC profile data: recipients, spending history, contribution feed
(port of lib/pacProfile.ts)."""

import asyncio
import logging
import math
import re

from ..config import fec_api_key
from ..core.cache import get_cache, set_cache
from ..core.db import (
    get_all_fec_candidates,
    get_members_from_db,
    get_pac_donations_by_name,
)
from ..core.http import shared_client
from ..core.util import iso_date, now_ms, parse_ms
from .pac_donations import normalize_party

logger = logging.getLogger("pac_profile")

RECIPIENTS_TTL_SECONDS = 6 * 60 * 60
COMMITTEE_ID_TTL_SECONDS = 30 * 24 * 60 * 60
COMMITTEE_MISS_TTL_SECONDS = 6 * 60 * 60
SPENDING_TTL_SECONDS = 12 * 60 * 60
REPORT_MAX_PAGES = 3

FEED_TTL_SECONDS = 12 * 60 * 60
FEED_MAX_PAGES = 15
FEED_LOOKBACK_MS = 5 * 365 * 24 * 60 * 60 * 1000


async def get_pac_recipients(pac_name: str) -> dict:
    cache_key = f"pac-recipients:{pac_name.lower()}"
    cached = await get_cache(cache_key)
    if cached:
        return cached

    rows, house, senate = await asyncio.gather(
        get_pac_donations_by_name(pac_name),
        get_members_from_db("house"),
        get_members_from_db("senate"),
    )

    by_id = {m["bioguide_id"]: m for m in [*house, *senate]}

    by_member: dict[str, dict] = {}
    for r in rows:
        member = by_id.get(r["bioguide_id"])
        if not member:
            continue
        try:
            amount = float(r["amount"])
        except (TypeError, ValueError):
            continue
        if not math.isfinite(amount) or amount <= 0:
            continue
        existing = by_member.get(r["bioguide_id"])
        if existing:
            existing["amount"] += amount
        else:
            by_member[r["bioguide_id"]] = {
                "bioguideId": r["bioguide_id"],
                "name": member["name"],
                "party": normalize_party(member["party"]),
                "chamber": "senate" if member["chamber"] == "senate" else "house",
                "amount": amount,
            }

    recipients = sorted(
        ({**r, "amount": round(r["amount"])} for r in by_member.values()),
        key=lambda r: r["amount"],
        reverse=True,
    )

    result = {
        "pacName": pac_name,
        "totalAmount": sum(r["amount"] for r in recipients),
        "houseCount": sum(1 for r in recipients if r["chamber"] == "house"),
        "senateCount": sum(1 for r in recipients if r["chamber"] == "senate"),
        "recipients": recipients,
    }

    if recipients:
        await set_cache(cache_key, result, RECIPIENTS_TTL_SECONDS)
    return result


_PAC_TYPE_PRIORITY = ["Q", "N", "O", "V", "W", "U", "D"]
_NAME_STOPWORDS = {
    "POLITICAL", "ACTION", "COMMITTEE", "FUND", "NATIONAL", "AMERICAN",
    "ASSOCIATION", "FEDERAL", "INC", "CORP", "CORPORATION", "COMPANY", "THE",
}


def _type_rank(committee_type: str | None) -> int:
    try:
        return _PAC_TYPE_PRIORITY.index(committee_type)
    except ValueError:
        return len(_PAC_TYPE_PRIORITY)


def _significant_tokens(name: str) -> list[str]:
    return [
        t
        for t in re.split(r"[^A-Z0-9]+", name.upper())
        if len(t) >= 4 and t not in _NAME_STOPWORDS
    ]


def _pick_committee(results: list[dict], tokens: list[str]) -> str | None:
    named = [c for c in results if c.get("committee_id") and c.get("name")]
    if tokens:
        matches = [
            c for c in named if any(t in c["name"].upper() for t in tokens)
        ]
    else:
        matches = named
    if not matches:
        return None
    matches.sort(key=lambda c: _type_rank(c.get("committee_type")))
    return matches[0].get("committee_id")


async def _search_committees(query: str, api_key: str) -> list[dict]:
    response = await shared_client().get(
        "https://api.open.fec.gov/v1/committees/",
        params={"api_key": api_key, "q": query, "per_page": "20"},
    )
    if response.status_code >= 400:
        return []
    return response.json().get("results") or []


async def _resolve_pac_committee_id(pac_name: str, api_key: str) -> str | None:
    cache_key = f"pac-committee-id:{pac_name.lower()}"
    cached = await get_cache(cache_key)
    if cached:
        return cached.get("id")

    cleaned = re.sub(r"\bPOLITICAL ACTION COMMITTEE\b", "", pac_name, flags=re.I)
    cleaned = re.sub(r"\bPAC\b", "", cleaned, flags=re.I)
    cleaned = re.sub(r"\s+", " ", cleaned).strip()
    queries = (
        [pac_name, cleaned]
        if cleaned and cleaned.upper() != pac_name.upper()
        else [pac_name]
    )
    tokens = _significant_tokens(pac_name)

    committee_id: str | None = None
    try:
        for query in queries:
            results = await _search_committees(query, api_key)
            committee_id = _pick_committee(results, tokens)
            if committee_id:
                break
    except Exception as error:
        logger.warning("resolve_pac_committee_id(%s) failed: %s", pac_name, error)

    await set_cache(
        cache_key,
        {"id": committee_id},
        COMMITTEE_ID_TTL_SECONDS if committee_id else COMMITTEE_MISS_TTL_SECONDS,
    )
    return committee_id


async def _fetch_pac_spending_history(committee_id: str, api_key: str) -> list[dict]:
    cache_key = f"pac-spending:{committee_id}"
    cached = await get_cache(cache_key)
    if cached:
        return cached

    reports: list[dict] = []
    for page in range(1, REPORT_MAX_PAGES + 1):
        response = await shared_client().get(
            f"https://api.open.fec.gov/v1/committee/{committee_id}/reports/",
            params={
                "api_key": api_key,
                "per_page": "100",
                "page": str(page),
                "sort": "-coverage_end_date",
            },
        )
        if response.status_code >= 400:
            break
        data = response.json()
        results = data.get("results") or []
        if not results:
            break
        reports.extend(results)
        pages = (data.get("pagination") or {}).get("pages") or 1
        if page >= pages:
            break

    by_period: dict[str, dict] = {}
    for r in reports:
        if not r.get("coverage_end_date") or r.get("most_recent") is False:
            continue
        day = r["coverage_end_date"][:10]
        existing = by_period.get(day)
        if not existing or (r.get("receipt_date") or "") > (existing.get("receipt_date") or ""):
            by_period[day] = r

    points = []
    for day, r in by_period.items():
        t = parse_ms(day)
        if t is None:
            continue
        try:
            disbursements = float(r.get("total_disbursements_period") or 0)
        except (TypeError, ValueError):
            disbursements = 0
        points.append({"t": t, "c": max(0, disbursements)})
    points.sort(key=lambda p: p["t"])

    if points:
        await set_cache(cache_key, points, SPENDING_TTL_SECONDS)
    return points


async def get_pac_spending(pac_name: str) -> dict:
    api_key = fec_api_key()
    if not api_key:
        return {"committeeId": None, "points": []}

    committee_id = await _resolve_pac_committee_id(pac_name, api_key)
    if not committee_id:
        return {"committeeId": None, "points": []}

    points = await _fetch_pac_spending_history(committee_id, api_key)
    return {"committeeId": committee_id, "points": points}


async def _fetch_pac_contributions(committee_id: str, api_key: str) -> list[dict]:
    min_date = iso_date(now_ms() - FEED_LOOKBACK_MS)
    out: list[dict] = []
    last_index: str | None = None
    last_date: str | None = None

    for _ in range(FEED_MAX_PAGES):
        params = {
            "api_key": api_key,
            "contributor_id": committee_id,
            "min_date": min_date,
            "per_page": "100",
            "sort": "-contribution_receipt_date",
        }
        if last_index:
            params["last_index"] = last_index
        if last_date:
            params["last_contribution_receipt_date"] = last_date

        response = await shared_client().get(
            "https://api.open.fec.gov/v1/schedules/schedule_a/", params=params
        )
        if response.status_code >= 400:
            break
        data = response.json()
        results = data.get("results") or []
        if not results:
            break

        for r in results:
            try:
                amount = float(r.get("contribution_receipt_amount") or 0)
            except (TypeError, ValueError):
                continue
            if not math.isfinite(amount) or amount <= 0:
                continue
            out.append(
                {
                    "recipientCommitteeId": r.get("committee_id"),
                    "candidateIds": (r.get("committee") or {}).get("candidate_ids") or [],
                    "date": r.get("contribution_receipt_date") or "",
                    "amount": amount,
                }
            )

        li = (data.get("pagination") or {}).get("last_indexes") or {}
        if not li.get("last_index") or li["last_index"] == last_index:
            break
        last_index = li["last_index"]
        last_date = li.get("last_contribution_receipt_date")

    return out


async def get_pac_contribution_feed(pac_name: str) -> dict:
    api_key = fec_api_key()
    if not api_key:
        return {"committeeId": None, "members": [], "contributions": []}

    committee_id = await _resolve_pac_committee_id(pac_name, api_key)
    if not committee_id:
        return {"committeeId": None, "members": [], "contributions": []}

    cache_key = f"pac-feed:{committee_id}"
    cached = await get_cache(cache_key)
    if cached:
        return cached

    raw, candidates, house, senate = await asyncio.gather(
        _fetch_pac_contributions(committee_id, api_key),
        get_all_fec_candidates(),
        get_members_from_db("house"),
        get_members_from_db("senate"),
    )

    committee_to_bioguide: dict[str, str] = {}
    candidate_to_bioguide: dict[str, str] = {}
    for c in candidates:
        for cid in c.get("committee_ids") or []:
            committee_to_bioguide[cid] = c["bioguide_id"]
        if c.get("candidate_id"):
            candidate_to_bioguide[c["candidate_id"]] = c["bioguide_id"]

    member_by_id = {m["bioguide_id"]: m for m in [*house, *senate]}

    contributions: list[dict] = []
    present: dict[str, dict] = {}
    for r in raw:
        bioguide = (
            committee_to_bioguide.get(r["recipientCommitteeId"])
            if r["recipientCommitteeId"]
            else None
        )
        if not bioguide:
            for cand in r["candidateIds"]:
                b = candidate_to_bioguide.get(cand)
                if b:
                    bioguide = b
                    break
        if not bioguide:
            continue
        member = member_by_id.get(bioguide)
        if not member:
            continue
        date = parse_ms(r["date"])
        if date is None:
            continue

        contributions.append(
            {"bioguideId": bioguide, "amount": round(r["amount"]), "date": date}
        )
        if bioguide not in present:
            present[bioguide] = {
                "bioguideId": bioguide,
                "name": member["name"],
                "party": normalize_party(member["party"]),
                "chamber": "senate" if member["chamber"] == "senate" else "house",
            }

    result = {
        "committeeId": committee_id,
        "members": list(present.values()),
        "contributions": contributions,
    }
    if contributions:
        await set_cache(cache_key, result, FEED_TTL_SECONDS)
    return result
